lib/spack/spack/package.py

```diff
@@ -949,7 +949,10 @@ def provides(self, vpkg_name):
         """
         True if this package provides a virtual package with the specified name
         """
-        return any(s.name == vpkg_name for s in self.provided)
+        return any(
+            any(self.spec.satisfies(c) for c in constraints)
```

Comment:
I think this makes more sense to be `all(...)` instead of `any(...)`, i.e. the constraints are combined with AND instead of OR. Other than that, this looks good. Thanks for the quick fix.

Reply:
Hmm, I thought about that, and OR seemed the right choice:
```python
provides('foo', when='+bar')
provides('foo', when='+baz')
```
With what I implemented a package provides `foo` if either `+bar` or `+baz` or both are satisfied. With what you suggest it will provide `foo` only if you have both `+bar+baz`. But for that you already have:
```python
provides('foo', when='+bar+baz')
```
I also think it's how other `when` parameters are combined for different directives (I am thinking of `depends_on` or `conflicts`, for instance). Do you see any fault in this reasoning?

Reply:
You are right. I was thinking of the case `when='+bar+baz'` and I though the `any(...)` applies to the individual parts: `+bar` OR `+baz`. However, I realized later that that is not the case and the different entries inside the `any(...)` command correspond to different `provides` commands, so using `any(...)` is the right thing.

One question: how are versions of virtual packages handled here? Does that require special treatment here? Should the check `s.name == vpkg_name` be replaced with something like `s.satisfies(vpkg_name)`, assuming that  `vpkg_name` may actually be a spec, not just a name?

Reply:
> how are versions of virtual packages handled here? Does that require special treatment here?

@v-dobrev Trying to reverse engineer the code, but I'd say no special treatment required.

`Spec.provides` is always called in contexts where a non-virtual package has been already chosen, what is taking care of the constraints on the virtual specs is the `ProviderIndex` which, given a virtual spec + constraints returns a list of non-virtual packages that could provide the requested service. 

Pinging @tgamblin so that he can correct me if I am wrong.